keps/672-execution-policy/README.md

```diff
@@ -0,0 +1,401 @@
+# KEP-672: Execution Policy
+
+<!--
+This is the title of your KEP. Keep it short, simple, and descriptive. A good
+title can help communicate what the KEP is and should be considered as part of
+any review.
+-->
+
+<!--
+A table of contents is helpful for quickly jumping to sections of a KEP and for
+highlighting any additional information provided beyond the standard KEP
+template.
+
+Ensure the TOC is wrapped with
+  <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code>
+tags, and then generate with `hack/update-toc.sh`.
+-->
+
+<!-- toc -->
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+    - [Story 3](#story-3)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [API Details](#api-details)
+  - [Implementation](#implementation)
+  - [Quota Management](#quota-management)
+  - [Defaulting/Validation](#defaultingvalidation)
+  - [Test Plan](#test-plan)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+  - [Manage quota for Job sequence](#manage-quota-for-job-sequence)
+  - [Support complex DAGs with JobSet](#support-complex-dags-with-jobset)
+- [Alternatives](#alternatives)
+  - [Add ExecutionPolicyRule parameter into the StartupPolicy API](#add-executionpolicyrule-parameter-into-the-startuppolicy-api)
+  <!-- /toc -->
+
+## Summary
+
+This KEP outlines the proposal to add the ExecutionPolicy API into JobSet. The ExecutionPolicy API
+allows to run sequence of ReplicatedJobs within a single JobSet.
+
+## Motivation
+
+Currently, JobSet supports the StartupPolicy API which allows to create Jobs in order after the
+first Job is in ready status. This can be useful when driver should be ready before workers.
+However, sometime high performance computing and machine learning users want to run sequence of
+ReplicatedJobs within JobSet. For example, it is common to run pre-processing,
+distributed fine-tuning, post-processing for LLM fine-tuning.
+
+### Goals
+
+- Add the ExecutionPolicy API for JobSet
+
+### Non-Goals
+
+- Support workflow management like DAGs outside of Job sequence.
+  - Users should consider to use Argo Workflows or Tekton Pipelines for such use-cases.
+- Allowing for a percentage of Jobs in a ReplicatedJob to be ready to consider the
+  whole ReplicatedJob to ready.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, I want to fine-tune LLM using multi-node training with JobSet. I have the first
```

Comment:
This story brings up some vagueness with the API. So you will start itiailizer and waiti for succeeded. Then I assume trainer job will start.

If I have a post-train replicated job, then do I also start the post-train replicated job once the initializer is complete.

We have executionPolicy InOrder but we don't specify what to do if the replicated job is not in the list.

Reply:
>If I have a post-train replicated job, then do I also start the post-train replicated job once the initializer is complete. We have executionPolicy InOrder but we don't specify what to do if the replicated job is not in the list.

If you have the post-train replicated job, it will start once the trainer job is complete, if you have this configuration:
```yaml
    - targetReplicatedJobs:
        - initializer
      waitForReplicatedJobsStatus: Succeeded
    - targetReplicatedJobs:
        - trainer-node
      waitForReplicatedJobsStatus: Succeeded
```
If you want to start trainer and post-processor job at the same time, after the initializer job is finished, just use this:
```yaml
    - targetReplicatedJobs:
        - initializer
      waitForReplicatedJobsStatus: Succeeded
```

Reply:
> If you want to start trainer and post-processor job at the same time, after the initializer job is finished, just use this:

This is a bit confusing to me because we only specify this with ExecutionPolicy: InOrder but if you don't specify any target replicated jobs than I guess it will start in any order?

Reply:
To support the InOrder policy, would having an explicit serial number (ID or similar) in the replicatedJobsSpec indicating the index of the jobs be more intuitive?

Reply:
>This is a bit confusing to me because we only specify this with ExecutionPolicy: InOrder but if you don't specify any target replicated jobs than I guess it will start in any order?

That is valid point. I think, if the InOrder execution policy is used, by default we should wait until ReplicatedJobs will be in the Ready status before create the next one. That will just replicate the existing StartupPolicy feature.
In that case, if executionPolicy is set for JobSet with 3 replicatedJobs the following settings will be used as a default one:
```yaml
executionPolicy:
  executionPolicyOrder: InOrder
  rules:
    - targetReplicatedJobs:
        - job-1
        - job-2
        - job-3
      waitForReplicatedJobsStatus: Ready
```

Any thoughts ?

Reply:
Can we list a few alternatives side-by-side to compare? 

I can also think of not having the rules at all in this struct, and have a `InOrderStartupRule` parameter in `ReplicatedJob` that takes values: `Succeeded|Ready`. This has the disadvantage of spreading the API though.

Reply:
Yes, this could be one of the options.
However, I tried to be consistent with other APIs, like [FailurePolicy](https://github.com/kubernetes-sigs/jobset/blob/main/api/jobset/v1alpha2/jobset_types.go#L313) or [SuccessPolicy](https://github.com/kubernetes-sigs/jobset/blob/main/api/jobset/v1alpha2/jobset_types.go#L325), where we indicate to which ReplicatedJobs we want to apply the rule.

@ahg-g Do you want me to add those API alternatives to the KEP ?